Add chainIds to revealed accounts in memberhsip requests

0x-r4bbit opened the pull request:
This commit does a few things:

- Adds a migration that adds chainids to `communities_request_to_join_revealed_addresses`
- Removes `RevealedAddress` in favor of `RevealedAccount` which is now a struct that contains the revealed address, as well as the signature and a list of chain IDs on which to check for user funds
- Changes the logic of sending requests to join a community, such that after creating address signatures, the user node will also check which of the addresses has funds on which networks for the community's token permissions, and add the chainds to the `RevealedAccount`
- Updates `checkPermissionToJoin()` such that only relevant chainids are used when checking user's funds. Chain IDs are retrieved from `RevealedAccount`s and matched against token permission criteria chain IDs

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (32)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 6061870a | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/1/) | 2023-05-16 09:39:50 | ~1 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/1/consoleText) |
| :heavy_multiplication_x: | 6061870a | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/1/) | 2023-05-16 09:45:26 | ~6 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/1//consoleText) |
| :x: | 6061870a | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/1/) | 2023-05-16 09:46:42 | ~8 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/1/consoleText) |
| :x: | 6061870a | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/1/) | 2023-05-16 09:46:42 | ~8 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/1/consoleText) |
| | | | | | | |
| :heavy_check_mark: | 03a04a93 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/2/) | 2023-05-16 10:42:24 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230516-103734-03a04a-pr3494.zip) |
| :heavy_multiplication_x: | 03a04a93 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/2/) | 2023-05-16 10:43:16 | ~5 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/2//consoleText) |
| :heavy_check_mark: | 03a04a93 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/2/) | 2023-05-16 10:44:26 | ~6 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230516-103738-03a04a-pr3494.zip) |
| :heavy_check_mark: | 03a04a93 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/2/) | 2023-05-16 10:46:41 | ~9 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230516-103734-03a04a-pr3494.aar) |
| | | | | | | |
| :heavy_check_mark: | e1c75832 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/3/) | 2023-06-06 12:54:04 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-125122-e1c758-pr3494.zip) |
| :heavy_multiplication_x: | e1c75832 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/3/) | 2023-06-06 12:54:28 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/3//consoleText) |
| :heavy_check_mark: | e1c75832 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/3/) | 2023-06-06 12:55:56 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-125122-e1c758-pr3494.aar) |
| :heavy_check_mark: | e1c75832 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/3/) | 2023-06-06 12:56:50 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-125122-e1c758-pr3494.zip) |
| | | | | | | |
| :heavy_multiplication_x: | 7c8de1c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/4/) | 2023-06-06 18:36:33 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/4//consoleText) |
| :heavy_check_mark: | 7c8de1c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/4/) | 2023-06-06 18:37:28 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-183517-7c8de1-pr3494.zip) |
| :heavy_multiplication_x: | 7c8de1c9 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/5/) | 2023-06-06 18:37:31 | ~53 sec | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/5//consoleText) |
| :heavy_check_mark: | 7c8de1c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/4/) | 2023-06-06 18:37:47 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-183513-7c8de1-pr3494.zip) |
| :heavy_check_mark: | 7c8de1c9 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/5/) | 2023-06-06 18:39:17 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-183732-7c8de1-pr3494.zip) |
| :heavy_check_mark: | 7c8de1c9 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/4/) | 2023-06-06 18:39:35 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-183513-7c8de1-pr3494.aar) |
| :heavy_check_mark: | 7c8de1c9 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/5/) | 2023-06-06 18:42:31 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-183939-7c8de1-pr3494.aar) |
| :heavy_check_mark: | 7c8de1c9 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/5/) | 2023-06-06 18:45:14 | ~7 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-183752-7c8de1-pr3494.zip) |
| | | | | | | |
| :heavy_check_mark: | f8af6c7b | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/6/) | 2023-06-06 19:03:53 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230606-190118-f8af6c-pr3494.zip) |
| :heavy_check_mark: | f8af6c7b | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/6/) | 2023-06-06 19:05:22 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230606-190118-f8af6c-pr3494.aar) |
| :heavy_check_mark: | f8af6c7b | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/6/) | 2023-06-06 19:07:25 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230606-190118-f8af6c-pr3494.zip) |
| :heavy_multiplication_x: | f8af6c7b | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/6/) | 2023-06-06 19:24:40 | ~23 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | c4b37f19 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/7/) | 2023-06-07 09:25:28 | ~9 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230607-091557-c4b37f-pr3494.zip) |
| :heavy_check_mark: | c4b37f19 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/7/) | 2023-06-07 09:35:33 | ~19 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230607-091557-c4b37f-pr3494.zip) |
| :heavy_check_mark: | c4b37f19 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/7/) | 2023-06-07 09:40:09 | ~24 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230607-091557-c4b37f-pr3494.aar) |
| :heavy_multiplication_x: | c4b37f19 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/7/) | 2023-06-07 09:58:46 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/7//consoleText) |
| :heavy_multiplication_x: | c4b37f19 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/8/) | 2023-06-07 10:48:40 | ~31 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/8//consoleText) |
| :heavy_multiplication_x: | c4b37f19 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/9/) | 2023-06-07 10:51:59 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/9//consoleText) |
| :heavy_multiplication_x: | c4b37f19 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/10/) | 2023-06-07 11:31:34 | ~27 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/10//consoleText) |
| :heavy_multiplication_x: | c4b37f19 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/11/) | 2023-06-07 14:02:25 | ~21 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/11//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 51533f9d | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/8/) | 2023-06-09 11:09:08 | ~1 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230609-110716-51533f-pr3494.aar) |
| :heavy_check_mark: | 51533f9d | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/8/) | 2023-06-09 11:10:17 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230609-110717-51533f-pr3494.zip) |
| :heavy_check_mark: | 51533f9d | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/8/) | 2023-06-09 11:10:25 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230609-110718-51533f-pr3494.zip) |
| :heavy_multiplication_x: | 51533f9d | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/12/) | 2023-06-09 11:21:48 | ~14 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/12//consoleText) |
| :heavy_check_mark: | 51533f9d | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/13/) | 2023-06-09 12:23:14 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/13//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 5060a1d2 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3494/9/) | 2023-06-12 07:02:24 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230612-065922-5060a1-pr3494.zip) |
| :heavy_check_mark: | 5060a1d2 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3494/9/) | 2023-06-12 07:04:17 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230612-065922-5060a1-pr3494.aar) |
| :heavy_check_mark: | 5060a1d2 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3494/9/) | 2023-06-12 07:05:39 | ~6 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230612-065922-5060a1-pr3494.zip) |
| :heavy_multiplication_x: | 5060a1d2 | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/14/) | 2023-06-12 07:21:16 | ~21 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/14//consoleText) |
| :heavy_multiplication_x: | 5060a1d2 | [#15](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/15/) | 2023-06-12 07:58:47 | ~21 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/15//consoleText) |
| :heavy_check_mark: | 5060a1d2 | [#16](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/16/) | 2023-06-12 08:05:19 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3494/16//consoleText) |

jrainville commented on protocol/communities/manager.go:
```diff
@@ -1501,25 +1487,90 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
+
+	ownedERC20TokenBalances := make(map[string]*big.Float)
+
+	if len(revealedChainIDsForERC20) > 0 {
```
Should we add a condition that we only get the balances if there is an ERC20 requirement (`len(erc20TokenRequirements) > 0` I guess?)

0x-r4bbit replied:
We already do that implicitly because `revealedChainIDsForERC20` have to exist in `erc20TokenRequirements`.
That's what we ensure in `getRevealedChainIDs()`.

If there's no chainIDs in the required permissions for ERC20 tokens, then `revealedChainIDsForERC20` will be empty

jrainville commented on protocol/communities/manager.go:
```diff
@@ -1501,25 +1487,90 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
+	}
+
+	ownedERC721Tokens := make(CollectiblesByChain)
+	if len(revealedChainIDsForERC721) > 0 {
```
Same thing here but for erc721

0x-r4bbit replied:
Same applies here.

jrainville commented on protocol/communities/manager.go:
```diff
@@ -1501,25 +1487,90 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
+	// we aren't able to check balances on selected networks
+	if len(erc20ChainIDsMap) > 0 && len(revealedChainIDsForERC20) == 0 {
+		return false, nil
+	}
```
Should we do the same check for erc721?

0x-r4bbit replied:
Actually we might even **remove** this check here.

Here's why:

The chain IDs we're checking on here are based on **all** token criteria from **all** permissions.
This means, it's perfectly fine to have criteria that need erc20 token balances, but not have any erc20 balances in an account **if** there's other criteria that, say, require ERC721 or ENS tokens which do exist in any of the user's accounts.

Remember that only **one** permission's criteria has to be met to be granted access.

So if you don't have any ERC20 tokens, even though there's a criteria in any of the permissions, no big deal, cause you might fulfill the criteria of a different permission.

Hence, we actually shouldn't return here.

jrainville reviewed (commented):
Great work. I add 3 comments inline about small possible improements

0x-r4bbit commented:
I'm working on refactoring this.
After discussing this with @cammellos we'll be piggy backing `checkPermissionsToJoin()` when sending a request to join.

0x-r4bbit committed:
Add chainIds to revealed accounts in memberhsip requests

This commit does a few things:

- Adds a migration that adds chainids to communities_request_to_join_revealed_addresses
- Removes RevealedAddress in favor of RevealedAccount which is now a struct that contains the revealed address, as well as the signature and a list of chain IDs on which to check for user funds
- Changes the logic of sending requests to join a community, such that after creating address signatures, the user node will also check which of the addresses has funds on which networks for the community's token permissions, and add the chainds to the RevealedAccount
- Updates checkPermissionToJoin() such that only relevant chainids are used when checking user's funds. Chain IDs are retrieved from RevealedAccounts and matched against token permission criteria chain IDs

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1313,7 +1317,13 @@ func (m *Manager) CheckPermissionToJoin(id []byte, addresses []gethcommon.Addres
+	}
+
+	accountsAndChainIDs := combineAddressesAndChainIDs(addresses, allChainIDs)
+	hasPermission, err := m.checkPermissionToJoin(becomeMemberPermissions, accountsAndChainIDs, false)
```
@cammellos here you can see the usage of `checkPermissionToJoin()` which now receives an accounts/chainID combination as discussed

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1543,14 +1548,11 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
-		}
-
-		permissionResponse, err := m.checkPermissionToJoin(becomeMemberPermissions, verifiedAddresses, true)
+		permissionResponse, err := m.checkPermissionToJoin(becomeMemberPermissions, accountsAndChainIDs, true)
```
@cammellos  Here's another usage where the accounts/chainIDs combination is received via the request-to-join

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1585,58 +1587,124 @@ func (m *Manager) HandleCommunityRequestToJoin(signer *ecdsa.PublicKey, request
+		if satisfied {
+			c.Satisfied = true
+		}
 	}
```
The above here turned out to be a logical bug.
As per spec, only one permission needs to be satisfied for someone to be able to join.
This fixes the logic such that this holds true.

Prior to this change, `Satisfied` would be set to `false` if just one token criteria from any permission wasn't.

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1793,50 +1919,23 @@ func (m *Manager) getOwnedERC721Tokens(walletAddresses []gethcommon.Address, tok
-	return ownedERC721Tokens, nil
-}
-
-func (m *Manager) getAccumulatedTokenBalances(accounts []gethcommon.Address, tokenRequirements map[uint64]map[string]*protobuf.TokenCriteria) (map[string]*big.Float, error) {
```
This function is no longer used as we now calculate accumulated balances on the fly

0x-r4bbit commented:
@cammellos  This needs some tests, but the basic idea is there

jrainville reviewed (approved):
Nice work!
